lib/os_places_api/client.rb

```diff
@@ -25,14 +25,23 @@ def locations_for_postcode(postcode, update: false)
 
       begin
         json = JSON.parse(response.body)
-        raise UnexpectedResponse if json["results"].nil?
 
-        if (existing = Postcode.find_by(postcode: postcode))
-          existing.update(results: json["results"], updated_at: Time.now) if update
+        if json["results"].nil?
+          # delete cached postcodes that were terminated by the Royal Mail
+          if (existing = Postcode.find_by(postcode: postcode))
+            existing.delete
+          else
+            raise UnexpectedResponse
+          end
+
         else
-          Postcode.create!(postcode: postcode, results: json["results"])
+          if (existing = Postcode.find_by(postcode: postcode))
+            existing.update(results: json["results"], updated_at: Time.now) if update
+          else
+            Postcode.create!(postcode: postcode, results: json["results"])
+          end
+          build_locations(json["results"])
```

Comment:
Rather than changing all this logic, I think we can just delete the postcode at the beginning of the method - see this spike: https://github.com/alphagov/locations-api/commit/68a0aa073d8d862da8f605cbde6c8598779ca48a

Makes it easier to follow the flow I think 👀

Reply:
Yes! This looks much better than the if...else saga